Config implementation

VermaShivansh opened the pull request.

VermaShivansh committed:
implementation of idea 1 as discussed in the docuemnt

VermaShivansh committed:
added testcases for testing ratelimit cofnig parser

VermaShivansh committed:
fixed logging and one of the testing function

VermaShivansh committed:
removed the error return from goroutines

VermaShivansh committed:
multizones are now supported and OR logic is used for ratelimit control

jcchavezs commented on plugin/plugin.go:
```diff
@@ -30,64 +32,33 @@ type Ratelimit struct {
-	interrupt     types.Interruption
+	ZoneMacro     macro.Macro
+	Action        string
+	Status        int // because coraza accepts 'int' in its interrupt struct
```
<3

jcchavezs commented on plugin/plugin.go:
```diff
@@ -133,12 +104,16 @@ func (e *Ratelimit) Evaluate(r rules.RuleMetadata, tx rules.TransactionState) {
 	if totalEventsOccuredInPreviousWindow < e.MaxEvents {
 		e.Zones[zone_name][currentTimeInSecond]++
-		fmt.Println(e.Zones)
+		log.Println(e.Zones)
```
Do you want to keep this?

VermaShivansh replied:
No in final release it should not be kept

jcchavezs commented on plugin/plugin.go:
```diff
@@ -133,12 +104,16 @@ func (e *Ratelimit) Evaluate(r rules.RuleMetadata, tx rules.TransactionState) {
 	} else {
 		// implement logic after ratelimit exceeded
-		fmt.Println("Ratelimit exceeded")
+		log.Println("Ratelimit exceeded")
```
Better to use the `tx` logger as it includes all the context from the transaction. See https://github.com/VermaShivansh/coraza-ratelimit-plugin/pull/8/files#diff-f42d64c75c9ca7a08c149ea609687764f2abcd3a79ae4e77110c6ae44ba36e9fR76

VermaShivansh replied:
sure, I thought I would use tx.Logger in the end for all logs once other things are done.
Since now it seems to be almost done, I will fix all loggers inside Evaluate to tx.Logger

jcchavezs commented on plugin/plugin.go:
```diff
@@ -162,6 +137,108 @@ func (e *Ratelimit) Type() rules.ActionType {
+	}
+
+	for _, token := range tokens {
+		pair := strings.Split(token, "=")
```
Probably better to use `strings.Cut`

VermaShivansh replied:
done and and added a test case for it as well

jcchavezs commented on plugin/plugin.go:
```diff
@@ -162,6 +137,108 @@ func (e *Ratelimit) Type() rules.ActionType {
+		switch key {
+		case "zone":
+			if e.ZoneMacro, err = macro.NewMacro(value); err != nil {
+				return fmt.Errorf("invalid macro name")
```
Use `errors.New` as you don't need to do interpolation

VermaShivansh replied:
done

jcchavezs commented on plugin/plugin.go:
```diff
@@ -162,6 +137,108 @@ func (e *Ratelimit) Type() rules.ActionType {
+			}
+			requiredValues[key] = true
+		case "events":
+			if e.MaxEvents, err = strconv.ParseInt(value, 10, 64); err != nil {
```
Maybe want to try `Atoi`

VermaShivansh replied:
the issue is Atoi converts to int, whereas I want int64, Although I can first convert to int and then convert it to int64 but that will require an extra variable and type casting. Whatever you suggest, should I change?

jcchavezs replied:
Makes sense.

jcchavezs commented on plugin/plugin.go:
```diff
@@ -162,6 +137,108 @@ func (e *Ratelimit) Type() rules.ActionType {
+			}
+			requiredValues[key] = true
+		case "interval":
+			if e.SweepInterval, err = strconv.ParseInt(value, 10, 64); err != nil {
```
I believe it is better to store this as `time.Duration`

VermaShivansh replied:
changing right away

VermaShivansh replied:
done

VermaShivansh committed:
changed logs inside evaluate to tx.Logger, removed and commented few loggers which will help in debugging application
using strings.Cut and added test for it
e.SweepInterval changed from int64 to time.Duration
using errors.New where interpolation is not happening

VermaShivansh committed:
lint fixes

VermaShivansh committed:
Merge pull request #9 from VermaShivansh/multizone

multizones are now supported and OR logic is used for ratelimit control

jcchavezs commented on default.conf:
```diff
@@ -6,5 +6,5 @@
 # SecRule REQUEST_BODY "@contains password" "id:3, phase:2,deny, status:403,msg:'Invalid request body',log,auditlog"
 
-SecRule ARGS:id "@eq 1" "id:1, ratelimit:zone=%{REQUEST_HEADERS.host}, pass, status:200"
+SecRule ARGS:id "@eq 1" "id:1, ratelimit:zone=%{REQUEST_HEADERS:host}&events=200&window=1&interval=1&action=drop&status=429, pass, status:200"
```
I'd rather do `zone[]` to make it explicit it is more than one.

jcchavezs commented on plugin/plugin.go:
```diff
@@ -27,67 +29,34 @@ type ZoneEvents map[int64]int64 // unixTimestamp in seconds containing requests
-	e.Window = 1
-	e.SweepInterval = 2
+	//default values
+	e.SweepInterval = 5
```
what means 5? can you do it in `time.Duration`?

jcchavezs commented on plugin/plugin.go:
```diff
@@ -104,46 +77,67 @@ func (e *Ratelimit) Evaluate(r rules.RuleMetadata, tx rules.TransactionState) {
-	currentTimeInSecond := time.Now().Unix()
+	// MultiZones behave in 'OR' manner, REQUEST will be allowed if any one of the zone is allowing the request
 
 	e.mutex.Lock()
```
since there is an iteration later I would move the lock down and do per iteration. Also work out the values like https://github.com/VermaShivansh/coraza-ratelimit-plugin/pull/8/files#diff-f42d64c75c9ca7a08c149ea609687764f2abcd3a79ae4e77110c6ae44ba36e9fR90-R94 out of the lock. Remember the more expensive the operation is, the more this becomes a bottleneck.

jcchavezs replied:
In a next PR we can explore RWMutex. For reference https://medium.com/bootdotdev/golang-mutexes-what-is-rwmutex-for-5360ab082626

jcchavezs commented on plugin/plugin.go:
```diff
@@ -152,19 +146,113 @@ func (e *Ratelimit) Evaluate(r rules.RuleMetadata, tx rules.TransactionState) {
+			requiredValues[key] = true
+		case "events":
+			if e.MaxEvents, err = strconv.ParseInt(value, 10, 64); err != nil {
+				return errors.New("invalid integer value for events")
```
Include the invalid value in the error to make it actionable.

VermaShivansh committed:
zone changed to zones, improved error handling, shifted mutex locking inside loop

VermaShivansh committed:
shifted code outside mutex lock

jcchavezs commented on plugin/constants.go:
```diff
@@ -0,0 +1,86 @@
+package plugin
```
I'd rather name this file `constants_test.go` as all this belongs to the tests.

jcchavezs commented on plugin/plugin.go:
```diff
@@ -152,19 +147,113 @@ func (e *Ratelimit) Evaluate(r rules.RuleMetadata, tx rules.TransactionState) {
+
+	for _, token := range tokens {
+		key, value, found := strings.Cut(token, "=")
+		if !found || key == "" || value == "" || strings.Count(value, "=") != 0 {
```
`strings.Count(value, "=")` will count all occurrences, while you want to know if there is one or not. `string.Contains` might do the trick.

VermaShivansh committed:
changed file name to constants_test.go and using strings.Contain instead of strings.Count to check for occurence of = in value
